source/common/unified_memory_allocation/include/uma/memory_pool.h

```diff
@@ -10,21 +10,42 @@
 #define UMA_MEMORY_POOL_H 1
 
 #include <uma/base.h>
-#include <uma/memory_pool_ops.h>
+#include <uma/memory_provider.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 typedef struct uma_memory_pool_t *uma_memory_pool_handle_t;
 
+struct uma_memory_pool_ops_t;
+
+/// \brief Type of a memory provider, defines purpose of the provider.
+enum uma_memory_provider_type {
+    UMA_MEMORY_PROVIDER_TYPE_METADATA,          ///< Provider should be used to allocate metadata only
+    UMA_MEMORY_PROVIDER_TYPE_DATA,              ///< Provider should be used to allocate data only
+    UMA_MEMORY_PROVIDER_TYPE_DATA_AND_METADATA, ///< Provider should be used to allocate both data and metadata
+    UMA_MEMORY_PROVIDER_TYPE_TRACKER            ///< Provider should not be used to request memory, it's only purpose is to track memory allocations.
```

Comment:
Again, it's mostly to support the case when pool does not use any memory provider for allocations.

> More generic question: do we need a tracking memory provider at all? Perhaps we can extend the umaMemoryProviderAlloc() function to track allocations?

We would need to extend `umaMemoryProviderAlloc` with a `pool` parameter so that the allocation can be associated with the correct pool. The approach with wrapping the memory provider is pretty generic and the only `ugly` interface is to support 'no memory provider' use case.